stdlib/dynarray.ml

```diff
@@ -102,17 +53,221 @@ and 'a slot =
+    - makes it explicit where dummies are used
+    - makes it hard to mistakenly mix data using distinct dummies,
+      which would be unsound *)
+module Dummy : sig
+
+  (** {4 Dummies} *)
+
+  type 'stamp dummy
+  (** The type of dummies is parametrized by a ['stamp] variable,
+      so that two dummies with different stamps cannot be confused
+      together. *)
+
+  type fresh_dummy = Fresh : 'stamp dummy -> fresh_dummy
+  val fresh : unit -> fresh_dummy
+  (** The type of [fresh] enforces a fresh/unknown/opaque stamp for
+      the returned dummy, distinct from all previous stamps. *)
+
+
+  (** {4 Values or dummies} *)
+
+  type ('a, 'stamp) with_dummy
+  (** a value of type [('a, 'stamp) with_dummy] is either a proper
+      value of type ['a] or a dummy with stamp ['stamp]. *)
+
+  val of_val : 'a -> ('a, 'stamp) with_dummy
+  val of_dummy : 'stamp dummy -> ('a, 'stamp) with_dummy
+
+  val is_dummy : ('a, 'stamp) with_dummy -> 'stamp dummy -> bool
+  val unsafe_get : ('a, 'stamp) with_dummy -> 'a
+  (** [unsafe_get v] can only be called safely if [is_dummy v dummy]
+      is [false].
+
+      We could instead provide
+      [val find : ('a, 'stamp) with_dummy -> ('a, 'stamp dummy) result]
+      but this would involve intermediary allocations.
+
+      {[match find x with
+        | None -> ...
+        | Some v -> ...]}
+      can instead be written
+      {[if Dummy.is_dummy x
+        then ...
+         else let v = Dummy.unsafe_get x in ...]}
+  *)
+
+  (** {4 Arrays of values or dummies} *)
+  module Array : sig
+    val make :
+      int -> 'a -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+
+    val init :
+      int -> (int -> 'a) -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+
+    val copy : 'a array -> dummy:'stamp dummy -> ('a, 'stamp) with_dummy array
+
+    val unsafe_nocopy :
+      'a array -> dummy:'stamp dummy ->
+      ('a, 'stamp) with_dummy array
+    (** [unsafe_nocopy] assumes that the input array was created
+        locally and will not be used anymore (in the spirit of
+        [Bytes.unsafe_to_string]), and avoids a copy of the input
+        array when possible. *)
+
+    val blit_array :
+      'a array -> int ->
+      ('a, 'stamp) with_dummy array -> int ->
+      len:int ->
+      unit
+    val prefix :
+      ('a, 'stamp) with_dummy array ->
+      int ->
+      ('a, 'stamp) with_dummy array
+    val extend :
+      ('a, 'stamp) with_dummy array ->
+      length:int ->
+      dummy:'stamp dummy ->
+      new_capacity:int ->
+      ('a, 'stamp) with_dummy array
+  end
+end = struct
+  type 'stamp dummy = { mutable dummy : 'stamp dummy }
+  [@@warning "-unused-field"] (* we never read the field *)
+  type fresh_dummy = Fresh : 'stamp dummy -> fresh_dummy
+
+  let fresh () =
+    (* dummy values and marshalling: we intentionally
+       use a cyclic value here. *)
+    let rec dummy = { dummy } in
+    Fresh dummy
+
+  type ('a, 'stamp) with_dummy = 'a
+
+  let of_val v = v
```

Comment:
I've been wondering if `let[@inline]` has an impact on this kind of code even without flambda. If it does, I think many of these functions should be annotated because we really don't want any overhead in dummy<->real value conversion.